### PR TITLE
fix: prevent a non-ALL role admin to assign ALL authority to role

### PR DESCRIPTION
**Pull request by netroms**

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java [hunk 1]
```diff
@@ -687,6 +685,27 @@ private void validateUserRoles(User user, User currentUser, List<ErrorReport> er
     }
   }
 
+  @Override
+  public List<ErrorReport> validateUserRoleCreateOrUpdate(UserRole role, User currentUser) {
```
Comment (vietnguyen) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java:689]: Do we need `  @Transactional(readOnly = true)` for this method ?

**Review by vietnguyen — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java [hunk 2]
```diff
@@ -687,6 +685,27 @@ private void validateUserRoles(User user, User currentUser, List<ErrorReport> er
...
+
+    if (currentUser == null || role == null) {
+      return errors;
+    }
+
+    if (!currentUser.isSuper() && role.isSuper()) {
+      errors.add(new ErrorReport(UserRole.class, ErrorCode.E3032, currentUser.getUsername()));
+    }
+
+    UserRole userRoleBefore = userRoleStore.get(role.getId());
```
Comment (david-mackessy) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java:701]: i don't follow why we go to the `userRoleStore` to get the same `userRole` that we already seem to have. Is there a specific reason for this?

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/SecurityCheck.java [hunk 3]
```diff
@@ -138,6 +139,17 @@ private <T extends IdentifiableObject> void runValidationCheck(
         }
       }
 
+      if (object instanceof UserRole) {
```
Comment (david-mackessy) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/SecurityCheck.java:142]: we could use the `instanceof` pattern match and remove the redundant cast e.g.
`object instanceof UserRole ur`

**Review by david-mackessy — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/SecurityCheck.java (hunk 3 above)
Comment (netroms) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/validation/SecurityCheck.java:142]: This PR needs to be backported to Java 11 branches, so for convenience, it's better to refactor this later.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java (hunk 2 above)
Comment (netroms) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java:701]: This is done to compare with the original unmodified role, the incoming role could be mutated.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java (hunk 1 above)
Comment (netroms) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/DefaultUserService.java:690]: Possibly, I added it to follow the same pattern as the other validation methods.

**Pull request merged**